common/include/pcl/impl/point_types.hpp

```diff
@@ -1828,6 +1835,561 @@ namespace pcl
+    (float, gradient_x, gradient_x)
+    (float, gradient_y, gradient_y)
+    (float, gradient_z, gradient_z)
+)
+
+POINT_CLOUD_REGISTER_POINT_STRUCT (pcl::PointWithScale,
+    (float, x, x)
+    (float, y, y)
+    (float, z, z)
+    (float, scale, scale)
+)
+
+POINT_CLOUD_REGISTER_POINT_STRUCT(pcl::PointSurfel,
+    (float, x, x)
+    (float, y, y)
+    (float, z, z)
+    (float, normal_x, normal_x)
+    (float, normal_y, normal_y)
+    (float, normal_z, normal_z)
+    (std::uint32_t, rgba, rgba)
+    (float, radius, radius)
+    (float, confidence, confidence)
+    (float, curvature, curvature)
+)
+
+POINT_CLOUD_REGISTER_POINT_STRUCT (pcl::_ReferenceFrame,
+    (float[3], x_axis, x_axis)
+    (float[3], y_axis, y_axis)
+    (float[3], z_axis, z_axis)
+)
+POINT_CLOUD_REGISTER_POINT_WRAPPER(pcl::ReferenceFrame, pcl::_ReferenceFrame)
+
+POINT_CLOUD_REGISTER_POINT_STRUCT (pcl::_PointDEM,
+    (float, x, x)
+    (float, y, y)
+    (float, z, z)
+    (float, intensity, intensity)
+    (float, intensity_variance, intensity_variance)
+    (float, height_variance, height_variance)
+)
+POINT_CLOUD_REGISTER_POINT_WRAPPER(pcl::PointDEM, pcl::_PointDEM)
 
+namespace pcl
+{
+
+// Allow float 'rgb' data to match to the newer uint32 'rgba' tag. This is so
+// you can load old 'rgb' PCD files into e.g. a PointCloud<PointXYZRGBA>.
+template<typename PointT>
+struct FieldMatches<PointT, ::pcl::fields::rgba>
+{
+  bool operator() (const pcl::PCLPointField& field)
+  {
+    if (field.name == "rgb")
+    {
+      // For fixing the alpha value bug #1141, the rgb field can also match
+      // uint32.
+      return ((field.datatype == pcl::PCLPointField::FLOAT32 ||
+               field.datatype == pcl::PCLPointField::UINT32) &&
+              field.count == 1);
+    }
+    else
+    {
+      return (field.name == traits::name<PointT, fields::rgba>::value &&
+              field.datatype == traits::datatype<PointT, fields::rgba>::value &&
+              field.count == traits::datatype<PointT, fields::rgba>::size);
+    }
+  }
+};
+template<typename PointT>
+struct FieldMatches<PointT, fields::rgb>
+{
+  bool operator() (const pcl::PCLPointField& field)
+  {
+    if (field.name == "rgba")
+    {
+      return (field.datatype == pcl::PCLPointField::UINT32 &&
+              field.count == 1);
+    }
+    else
+    {
+      // For fixing the alpha value bug #1141, rgb can also match uint32
+      return (field.name == traits::name<PointT, fields::rgb>::value &&
+              (field.datatype == traits::datatype<PointT, fields::rgb>::value ||
+               field.datatype == pcl::PCLPointField::UINT32) &&
+              field.count == traits::datatype<PointT, fields::rgb>::size);
+    }
+  }
+};
+
+
+// We're doing a lot of black magic with Boost here, so disable warnings in Maintainer mode, as we will never
+// be able to fix them anyway
+#if defined _MSC_VER
+  #pragma warning(disable: 4201)
 #endif
+
+namespace traits
```

Comment:
2400 line file is trouble. We definitely need to split this (in another PR). Maybe if we can make some things point independent then we can add some to `type_traits.h` since that's what it is from this point onward.

Note for future: It's important that `type_traits.h` is type independent.